strawberryfields/engine.py

```diff
@@ -182,58 +285,34 @@ def _run_program_locally(self, prog, **kwargs):
-        """Execute the given program by sending it to the backend.
+    def run(self, program, *, compile=True, modes=None, state_options={}, **kwargs):
+        """Execute the given programs by sending them to the backend.
 
-        * The backend state is updated.
-        * The executed Program(s) are appended to self.run_progs.
+        Extends :meth:`BaseEngine._run`.
 
         Args:
-            program (Program, Sequence[Program]): quantum circuit(s) to run
-            return_state (bool): If True, returns the state of the circuit after the
-                circuit has been run like :meth:`return_state` was called.
-            modes (Sequence[int]): Modes to be returned in the state object. If None, returns all modes.
-            compile (bool): If True, compile the Program instances before sending them to the backend.
-        """
-        if not isinstance(program, Sequence):
-            program = [program]
-
-        # TODO unsuccessful run due to exceptions should ideally have no effect on the backend state (state checkpoints?)
-        try:
-            prev = None  # previous program segment
-            for p in program:
-                if self.run_progs:
-                    prev = self.run_progs[-1]
-                else:
-                    # initialize the backend TODO where should this happen? c.f. the fixtures in conftest.py
-                    self.backend.begin_circuit(num_subsystems=p.init_num_subsystems, **self.kwargs)
+            modes (None, Sequence[int]): Modes to be returned in the ``Result.state`` :class:`.BaseState` object.
+                An empty sequence [] means no state object is returned. None returns all the modes.
+            state_options (Dict[str, Any]): keyword arguments for :meth:`.BaseBackend.state`
 
-                if prev is not None and not p.can_follow(prev):
-                    raise RuntimeError("Register mismatch after Program {}, '{}'.".format(len(self.run_progs)-1, prev.name))
-
-                # if the program hasn't been compiled for this backend, do it now
-                if compile and p.backend != self.backend._short_name:
-                    p = p.compile(self.backend._short_name)
-                p.lock()
+        Returns:
+            Result: results of the computation
+        """
 
-                # TODO handle remote backends here, store measurement results in the RegRefs or maybe in the Engine object.
-                self._run_program_locally(p, **kwargs)
-                self.run_progs.append(p)
-        except Exception as e:
-            # TODO reset the backend to the last checkpoint here?
-            raise e
-        else:
-            # program execution was successful
+        result = super()._run(program, compile=compile, **kwargs)
+        if isinstance(modes, Sequence) and not modes:
+            # empty sequence
             pass
+        else:
+            result.state = self.backend.state(modes, **state_options)  # tfbackend.state can use kwargs
```

Comment:
Oh, `state_options` is for the TF backend only?

Reply:
Yeah, I think so.